browser/base/content/newtab/newTab.js

```diff
@@ -35,6 +35,68 @@ function inPrivateBrowsingMode() {
   return PrivateBrowsingUtils.isContentWindowPrivate(window);
 }
 
+function handleCommand(command, data) {
+  let commandHandled = true;
+  switch(command) {
+    case "NewTab:UpdateTelemetryProbe":
+      Services.telemetry.getHistogramById(data.probe).add(data.value);
+      break;
+    case "NewTab:Register":
+      registerEvent(data.type);
+      break;
+    case "NewTab:GetInitialState":
+      getInitialState();
+      break;
+    default:
+      commandHandled = false;
+  }
+  return commandHandled;
+}
+
+function initRemotePage() {
+  // Messages that the iframe sends the browser will be passed onto
+  // the privileged parent process
+  let iframe = document.getElementById("meep");
+  iframe.addEventListener("load", (e) => {
+    iframe.contentDocument.addEventListener("NewTabCommand", (e) => {
+      let handled = handleCommand(e.detail.command, e.detail.data);
+      if (!handled) {
+        sendAsyncMessage(e.detail.command, e.detail.data);
+      }
+    }, false);
+    registerEvent("NewTab:Observe");
+    iframe.contentDocument.dispatchEvent(new CustomEvent("NewTabCommandReady"));
+  });
+}
+
+function registerEvent(event) {
+  // Messages that the privileged parent process sends will be passed
+  // onto the iframe
+  addMessageListener(event, (message) => {
+    let iframe = document.getElementById("meep").contentWindow;
+    iframe.postMessage(message, "*");
+  });
+}
+
+function getInitialState() {
+  let state = {
+    enabled: Services.prefs.getBoolPref(PREF_NEWTAB_ENABLED),
+    enhanced: Services.prefs.getBoolPref(PREF_NEWTAB_ENHANCED),
+    rows: Services.prefs.getIntPref(PREF_NEWTAB_ROWS),
+    columns: Services.prefs.getIntPref(PREF_NEWTAB_COLUMNS),
+    introShown: Services.prefs.getBoolPref(PREF_INTRO_SHOWN),
+    privateBrowsingMode: PrivateBrowsingUtils.isContentWindowPrivate(window)
+  }
+  let iframe = document.getElementById("meep").contentWindow;
+  iframe.postMessage({name: "NewTab:State", data: state}, "*");
+}
+
+const PREF_NEWTAB_ENABLED = "browser.newtabpage.enabled";
```

Comment:
Could we move all these to the top of the file? Just convention to have static constant there.